### DISCUSSION TITLE
[Q]data validation for Time

### DISCUSSION THREAD

**Discussion by ayushkumar9333** [exceljs/exceljs#2448] [Q&A]
i have two fields in my dummy project one of date and other of time.

for date i an setting custom validation as 
 ****if (cell.address == 'E1') {
            var E = 'E' + (a);
            worksheet.getCell(E).dataValidation = {
              type: 'date',
              operator: 'greaterThanOrEqual',
              formulae: [new Date()],
              showInputMessage: true,
              allowBlank: true,
              promptTitle: 'Port Date ',
              prompt: 'Format  (MM/DD/YYYY)',
              showErrorMessage: true,
              errorStyle: 'error',
              errorTitle: 'Port Date ',
              error: 'Format  (MM/DD/YYYY)',

            };
            worksheet.getCell(E).numFmt = 'mm/dd/yyyy';
            worksheet.getCell(E).alignment = { vertical: 'middle', horizontal: 'center' };
          }****

i want to use it for time in h:mm:ss AM/PM and set my validation between Start time 12:00:00 AM 
and End time 11:59:59 PM 

          if (cell.address == 'F1') {
            var F = 'F' + (a);
            worksheet.getCell(F).dataValidation = {
              type: 'date',
              operator: 'between',
              formulae: [new Date(12,31,2019),new Date(12,31,2019)],
              //formulae2: ['TEXT(TIME(23,18,14),"h:mm:ss AM/PM")'],
              showInputMessage: true,
              allowBlank: true,
              promptTitle: 'Port Time ',
              prompt: 'Format (HR:MIN:SS AM/PM)',
              showErrorMessage: true,
              errorStyle: 'error',
              errorTitle: 'Port Time ',
              error: 'Format (HR:MIN:SS AM/PM)',

            };
            worksheet.getCell(F).numFmt = 'h:mm:ss AM/PM';
            worksheet.getCell(F).alignment = { vertical: 'middle', horizontal: 'center' };
          }
Please help me out with using Time

**Comment by Shravan1106** [comment]
I am working on a project where I need similar data validation for time but for 24-hr format.Did you solve the above issue..?

**Comment by MigsSalazar** [comment]
I was able to find a bit of a hack that isn't technically supported but it IS how the excel works. 
```
{
    type: 'time',
    allowBlank: true,
    operator: 'between',
    formulae: [0, 0.999988425925926],
    //any other standard props you need
}
```

I got to this answer by reading in an excel workbook with the desired data validation settings on a cell and printing the cell's read in data validation object to console. For full transparency, the formulae I got back was actually `formulae: ["0", "0.999988425925926"]` so the numbers were in strings, but the validation worked with numbers. 

Using `1` as the second formulae parameter DOES NOT WORK. I haven't tested any other values besides 1 and `0.999988425925926`, but this longer decimal has worked for me.

If you are using Typescript (as I am) you can cast it with `cell.dataValidation = {this is your object} as any as DataValidation`

**Comment by MigsSalazar** [comment]
Quick little update for TypeScript users, I made a monkey patch `.d.ts` file so that I didn't have to convert to any then DataValidation. To truly add it in, add the following in a `XLSX.d.ts` file (or whatever name you want as long as it's `.d.ts` somewhere in your project

```
import "exceljs"

declare module "exceljs" {
	export interface DataValidationWithTime extends DataValidation {
		type: 'time' | DataValidation['type'];
	}

	export interface DataValidations {
		add(range: string, validation: DataValidationWithTime): void;
		readonly model: Record<string, DataValidationWithTime>;
	}

	export interface Worksheet {
		dataValidations: DataValidations;
	}
};
```
